lib/guest_installation_and_configuration_base.pm

```diff
@@ -0,0 +1,1581 @@
+        type_string("reset\n");
+        wait_still_screen;
+    }
+    else {
+        record_info("Failed to detach $self->{guest_name} installation screen process $self->{guest_installation_session}", "Bad luck !");
+        reset_consoles;
+        select_console('root-ssh');
+        $self->get_guest_installation_session if ($self->{guest_installation_session} eq '');
+        type_string("reset\n");
+        wait_still_screen;
+    }
+    $self->{guest_installation_attached} = 'false';
+    return $self;
+}
+
+#Record final guest installation result in [guest_installation_result] and set [stop_run] and [stop_timestamp].
+sub record_guest_installation_result {
+    my ($self, $_guest_installation_result) = @_;
+
+    $self->reveal_myself;
+    $_guest_installation_result //= '';
+    $self->{guest_installation_result} = $_guest_installation_result;
+    record_info("Guest $self->{guest_name} installation has already been marked as $self->{guest_installation_result}", "It is done !");
+    $self->{stop_run}       = time();
+    $self->{stop_timestamp} = localtime($self->{stop_run});
+    return $self;
+}
+
+#Collect guest y2logs via ssh and save guest config xml file.
+sub collect_guest_installation_logs_via_ssh {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->get_guest_ipaddr;
+    if ((script_run("nmap $self->{guest_ipaddr} -PN -p ssh | grep -i open") eq 0) and ($self->{guest_ipaddr} ne '') and ($self->{guest_ipaddr} ne 'NO_IP_ADDRESS_FOUND_AT_THE_MOMENT')) {
+        record_info("Guest $self->{guest_name} has ssh port open on ip address $self->{guest_ipaddr}.", "Try to collect logs via ssh but may fail.Open ssh port does not mean good ssh connection.");
+        script_run("ssh -vvv root\@$self->{guest_ipaddr} \"save_y2logs /tmp/$self->{guest_name}_y2logs.tar.gz\"");
+        script_run("scp -r -vvv -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$self->{guest_ipaddr}:/tmp/$self->{guest_name}_y2logs.tar.gz $self->{guest_log_folder}");
+    }
+    else {
+        record_info("Guest $self->{guest_name} has no ssh connection available at all.Not able to collect logs from it via ssh", "Guest ip address is $self->{guest_ipaddr}");
+    }
+    script_run("virsh dumpxml $self->{guest_name} > $self->{guest_log_folder}/virsh_dumpxml_$self->{guest_name}.xml");
+    script_run("rm -f -r $common_log_folder/autoyast_*");
+    return $self;
+}
+
+#Upload logs collect by collect_guest_installation_logs_via_ssh.
+sub upload_guest_installation_logs {
+    my $self = shift;
+
+    $self->reveal_myself;
+    assert_script_run("tar czvf /tmp/guest_installation_and_configuration_logs.tar.gz $common_log_folder");
+    upload_logs("/tmp/guest_installation_and_configuration_logs.tar.gz");
+    return $self;
+}
+
+#Power cycle guest by force:virsh destroy,grace:virsh shutdown,reboot:virsh reboot and poweron:virsh start.
+sub power_cycle_guest {
```

Comment:
As the same, would be better to use with assert_script_run() to admin guest under power_cycle_guest()

Reply:
assert_scrip_run is not very suitable here, because the power_cycle_guest here is to facilitate guest installation, especially those with "noautoconsole" and etc. So such guests may be in shutdown state at certain stage which needs to be powered on again. But you can not tell exactly whether the guest involved is really in shutdown state when perform the operation due to timing issue or similar. The guest may change to shutdown state after the moment when you see it in running state and vice versa. So we just need to ensure it is powered on again or put them into other power state. It is predictable that assert_script_run will cause many unnecessary operations failures which also leads to test failure which is totally unnecessary.

Reply:
yes, totally make sense now. keep using `script_run() `here for `power_cycle_guest()`.